### PR TITLE
Adds support for ML models

### PR DESCRIPTION
**Pull request by nopper**
Hi there,

this PR adds support for deploying and querying custom trained ML models using the [Vertex AI platform](https://cloud.google.com/vertex-ai).

The main contributions are two bash scripts and a python library:
 - `setup_vpc.sh` which sets up the VPC peering allowing to deploy private endpoints in the current infrastructure.
 - `deploy_ai.sh` which deploys a model reading its name and bucket from a yaml file.
 - `ai.py` python library that offers an `InferenceClient` object for interacting with a given model, using different protocols (rest or grpc supported).
 
The model is configured using the `ai.yaml` file. In there you need specify the `endpoint_id` and `deployed_model_id`. This last attribute is optional, as public endpoints over rest api don't need this parameter.

Couple of notes:
1. This can be generalized to multi-model deployment pretty easily.
2. I modified the search interface just to demo that this is working fine. Do you have any suggestions regarding a more thorough testing strategy?

### CODE REVIEW

File: deploy/gke/dev.yaml [hunk 1]
```diff
@@ -21,3 +21,7 @@ region:
     - europe-west2
 nodes: 1
 storage_project: datcom-store
+# Network name used for VPC peerings.
+network_name: "default"
+# Assumes a directory with this name exists in gs://datcom-website-dev-modelstore/
```
Comment (shifucun) [inline deploy/gke/dev.yaml:26]: I guess you also want a field for the model bucket?

File: deploy/overlays/prod/kustomization.yaml [hunk 2]
```diff
@@ -40,6 +40,9 @@ configMapGenerator:
   - name: redis-config
     files:
       - redis.json
+  - name: ai-config
```
Comment (shifucun) [inline deploy/overlays/prod/kustomization.yaml:43]: Instead of "prod" maybe can first put these in "autopush" folder, which is an instance used for testing.

File: scripts/setup_nlp.sh [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+if [[ $REGION == "" ]]; then
+  REGION=$(yq eval '.region.primary' $ROOT/deploy/gke/$ENV.yaml)
+fi
+
+PROJECT_ID=$(yq eval '.project' $ROOT/deploy/gke/$ENV.yaml)
+PROJECT_NUMBER=$(gcloud projects list --filter="${PROJECT_ID}" --format="value(PROJECT_NUMBER)")
+
+MODEL_NAME=$(yq eval '.model_name' $ROOT/deploy/gke/$ENV.yaml)
+NETWORK_NAME=$(yq eval '.network_name' $ROOT/deploy/gke/$ENV.yaml)
```
Comment (shifucun) [inline scripts/setup_nlp.sh:37]: maybe we should just use "default" for NETWORK_NAME? Also can remove it from yaml file until we need this to be a variable.

File: scripts/setup_nlp.sh [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+MODEL_BUCKET="gs://${PROJECT_ID}-modelstore"
```
Comment (shifucun) [inline scripts/setup_nlp.sh:39]: MODEL_BUCKET can be moved to yaml

File: scripts/setup_nlp.sh [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+MODEL_PATH="${MODEL_BUCKET}/${MODEL_NAME}"
+
+if ! gsutil -q stat "${MODEL_PATH}/*" ; then
+  echo "Model path does not exists ${MODEL_PATH}"
+  
+  # This might fail, meaning that the bucket already exists.
```
Comment (shifucun) [inline scripts/setup_nlp.sh:45]: If bucket exists, it can not get through the "if" check?

File: scripts/setup_nlp.sh [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+gcloud beta ai endpoints create \
+    --project="${PROJECT_ID}" \
+    --region="${REGION}" \
+    --network="projects/${PROJECT_NUMBER}/global/networks/${NETWORK_NAME}" \
+    --display-name="${VERTEX_ENDPOINT_NAME}"
+
+ENDPOINT_ID=$(gcloud ai endpoints list --project="${PROJECT_ID}" --region="${REGION}" --filter "display_name=${VERTEX_ENDPOINT_NAME}" --format="value(ENDPOINT_ID.scope())")
+MODEL_ID=$(gcloud ai models list --project="${PROJECT_ID}" --region="${REGION}" --filter "display_name=${VERTEX_MODEL_NAME}" --format="value(MODEL_ID.scope())")
+
+gcloud beta ai endpoints deploy-model "${ENDPOINT_ID}" \
```
Comment (shifucun) [inline scripts/setup_nlp.sh:77]: Does this script contains one time setup or it is used for each new deployment. If it has one time setup, should move it to $ROOT/gke as "setup_nlp.sh". Then the remaining part can be renamed as "deploy_nlp.sh" since it's not really a setup?

File: scripts/setup_vpc.sh [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+# limitations under the License.
+
+# This scripts sets up a VPC peering connection to servicenetworking.
+
+set -e
+
+ENV=$1
+REGION=$2
+
+if [[ $ENV != "staging" && $ENV != "prod" && $ENV != "autopush" && $ENV != "dev" && $ENV != "feeding-america" && $ENV != "tidal" && $ENV != "karnataka" ]]; then
```
Comment (shifucun) [inline scripts/setup_vpc.sh:23]: I think this is one time setup. Can move to $ROOT/gke/setup_vpc.sh

File: server/configmodule.py [hunk 8]
```diff
@@ -102,6 +103,8 @@ class LocalConfig(Config):
     SECRET_PROJECT = 'datcom-website-dev'
     API_ROOT = 'https://autopush.api.datacommons.org'
     RECON_API_ROOT = 'https://autopush.recon.datacommons.org'
+    AI_CONFIG_PATH = os.path.abspath(
```
Comment (shifucun) [inline server/configmodule.py:106]: You can mount it in the GKE. An example is redis.json https://github.com/datacommonsorg/website/blob/master/deploy/overlays/prod/patch_deployment.yaml#L38. Then you don't have do this for each environment as below

File: server/routes/search.py [hunk 9]
```diff
@@ -45,8 +46,10 @@ def search_dc():
     else:
         search_response = {}
 
-    # Convert from search results to template dictionary.
     results = []
+    results.append(ai.search(query_text))
```
Comment (shifucun) [inline server/routes/search.py:50]: Can create a new route like '/search_nlp' which only has result from this?

File: server/services/ai.py [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+if not cfg.TEST and cfg.AI_CONFIG_PATH:
+    _INFERENCE_CLIENT = create_inference_client(cfg.AI_CONFIG_PATH)
+
+
+def search(query: str) -> Optional[Any]:
+    global _INFERENCE_CLIENT
+
+    if _INFERENCE_CLIENT:
+        response = _INFERENCE_CLIENT.request(query)
+        # This is a fake response for the moment.
```
Comment (shifucun) [inline server/services/ai.py:179]: To confirm, does this return real inference for dev instance?

**Comment by shifucun** [comment]
Thanks for sending this. Overall looks very good and structured! Just some minor comments

File: scripts/setup_nlp.sh (hunk 3 above)
Comment (nopper) [reply scripts/setup_nlp.sh:37]: Done

Comment (nopper) [reply scripts/setup_nlp.sh:37]: Done.

File: scripts/setup_nlp.sh (hunk 4 above)
Comment (nopper) [reply scripts/setup_nlp.sh:39]: Done.

File: scripts/setup_nlp.sh (hunk 5 above)
Comment (nopper) [reply scripts/setup_nlp.sh:45]: This checks if a model path exists in the specified bucket. If the bucket exists but a model is not there this will cause the script to exit.

I removed the bucket creation in case this fails as it might not be useful. It should be the user responsibility I guess to make sure everything is in check.

File: scripts/setup_nlp.sh (hunk 6 above)
Comment (nopper) [reply scripts/setup_nlp.sh:77]: Moved it to `deploy_ai.sh` which is a more generic name.

File: scripts/setup_vpc.sh (hunk 7 above)
Comment (nopper) [reply scripts/setup_vpc.sh:23]: Done.

File: deploy/gke/dev.yaml (hunk 1 above)
Comment (nopper) [reply deploy/gke/dev.yaml:26]: Done.

File: server/services/ai.py (hunk 10 above)
Comment (nopper) [reply server/services/ai.py:179]: Yes. I tested it using the public endpoint. But I also like to test it on a dev environment to make sure the grpc client works as expected.

Do you have any instructions on how to do that?

File: server/routes/search.py (hunk 9 above)
Comment (nopper) [reply server/routes/search.py:50]: Done.

File: server/services/ai.py (hunk 10 above)
Comment (nopper) [reply server/services/ai.py]: Tested on dev. Seems to work fine.

**Comment by nopper** [comment]
PTAL. I've addressed all your comments and tested it on dev.

File: deploy/overlays/autopush/ai.yaml [hunk 11]
```diff
@@ -0,0 +1,4 @@
+us-central1:
```
Comment (shifucun) [inline deploy/overlays/autopush/ai.yaml:1]: I don't see model deployed in datcom-website-autopush?

File: deploy/overlays/dev/kustomization.yaml [hunk 12]
```diff
@@ -50,3 +53,6 @@ patchesStrategicMerge:
           maxSurge: 2
           # maximum number of Pods that can be unavailable during the update process
           maxUnavailable: 75%
+
+patchesStrategicMerge:
```
Comment (shifucun) [inline deploy/overlays/dev/kustomization.yaml:57]: Can you move line 41-54 to patch_deployment.yaml too?

File: deploy/overlays/dev/patch_deployment.yaml [hunk 13]
```diff
@@ -0,0 +1,36 @@
+# Copyright 2019 Google LLC
```
Comment (shifucun) [inline deploy/overlays/dev/patch_deployment.yaml:1]: 2022

File: deploy/overlays/prod/patch_deployment.yaml [hunk 14]
```diff
@@ -31,8 +31,13 @@ spec:
         - name: redis-config
           configMap:
             name: redis-config
+        - name: ai-config
```
Comment (shifucun) [inline deploy/overlays/prod/patch_deployment.yaml:34]: Maybe should move this to $ROOT/deploy/base/deployment.yaml? So this will be a shared pattern. Then a lot of patch above can be saved.

File: gke/setup_vpc.sh [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+set -e
+
+ENV=$1
+REGION=$2
+
+if [[ $ENV != "staging" && $ENV != "prod" && $ENV != "autopush" && $ENV != "dev" && $ENV != "feeding-america" && $ENV != "tidal" && $ENV != "karnataka" ]]; then
+  echo "First argument should be 'staging' or 'prod' or 'autopush' or 'dev' or 'feeding-america' or 'tidal' or 'karnataka'"
+  exit
+fi
```
Comment (shifucun) [inline gke/setup_vpc.sh:26]: I have removed this check in other scripts too, since this is not scalable.. Can remove there.

**Review by shifucun — COMMENTED** [review]
Can you revert the change for mixer submodule update?

File: scripts/deploy_ai.sh [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+set -e
+
+ENV=$1
+REGION=$2
+
+if [[ $ENV != "staging" && $ENV != "prod" && $ENV != "autopush" && $ENV != "dev" && $ENV != "feeding-america" && $ENV != "tidal" && $ENV != "karnataka" ]]; then
+  echo "First argument should be 'staging' or 'prod' or 'autopush' or 'dev' or 'feeding-america' or 'tidal' or 'karnataka'"
+  exit
+fi
+
```
Comment (shifucun) [inline scripts/deploy_ai.sh:25]: Can remove the check here...

File: server/configmodule.py (hunk 8 above)
Comment (shifucun) [reply server/configmodule.py:111]: Got it, that makes sense.

**Review by shifucun — COMMENTED** [review]
Overall looks good! Just some nit about the ai.yaml config change. Really appreciate sending the PR and testing it throughout!

File: gke/setup_vpc.sh (hunk 15 above)
Comment (nopper) [reply gke/setup_vpc.sh:26]: Done.

File: deploy/overlays/autopush/ai.yaml (hunk 11 above)
Comment (nopper) [reply deploy/overlays/autopush/ai.yaml]: Changed with an empty file for now.

**Comment by nopper** [comment]
Hi @shifucun . Addressed your comments. PTAL.

File: deploy/git/mixer_hash.txt [hunk 17]
```diff
@@ -0,0 +1 @@
+be02c54
```
Comment (shifucun) [inline deploy/git/mixer_hash.txt:1]: please revert this change

File: deploy/git/website_hash.txt [hunk 18]
```diff
@@ -0,0 +1 @@
+5d47748
```
Comment (shifucun) [inline deploy/git/website_hash.txt:1]: please revert this change

File: deploy/overlays/dev/kustomization.yaml [hunk 19]
```diff
@@ -34,6 +34,9 @@ configMapGenerator:
     literals:
       - mixerProject=datcom-website-dev
       - serviceName=website-esp.endpoints.datcom-website-dev.cloud.goog
+  - name: ai-config
+    files:
+      - ai.yaml
```
Comment (shifucun) [inline deploy/overlays/dev/kustomization.yaml:39]: can move this to base/kustomization.yaml as well and define empty ai.yaml for all env that has no model deployed

File: deploy/overlays/prod/patch_deployment.yaml (hunk 14 above)
Comment (shifucun) [inline deploy/overlays/prod/patch_deployment.yaml:34]: This can be removed since it's added to base

File: deploy/overlays/prod/patch_deployment.yaml [hunk 20]
```diff
@@ -31,8 +31,13 @@ spec:
...
+          configMap:
+            name: ai-config
       containers:
         - name: website
           volumeMounts:
             - name: redis-config
               mountPath: /datacommons/redis
+            - name: ai-config
```
Comment (shifucun) [inline deploy/overlays/prod/patch_deployment.yaml:42]: can be removed as well

**Review by shifucun — COMMENTED** [review]
The mixer submodule is still not reverted. Please revert.

File: deploy/overlays/prod/patch_deployment.yaml (hunk 14 above)
Comment (nopper) [reply deploy/overlays/prod/patch_deployment.yaml:34]: Done.

File: deploy/overlays/dev/kustomization.yaml (hunk 19 above)
Comment (nopper) [reply deploy/overlays/dev/kustomization.yaml:39]: Resolved offline.

**Review by shifucun — APPROVED** [review]
Thanks for making all the update!
